### PR TITLE
[skip changelog] Update contributions guidelines and PR template to add labeling and increase focus on breaking changes

### PR DESCRIPTION
**Pull request by rsora**
**Please check if the PR fulfills these requirements**

- [x] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [x] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?**

This PR updates the Contribution guidelines and the PR template to ask contributors and maintainers to use the label **`[breaking]`** when a breaking change is introduced. 
Our internal release procedures are updated accordingly as well in order to:
      - Extend the QA window we have between the release candidate creation and the actual release.
      - Highlight better breaking changes in the changelog if any
      - Created a process to handle breaking changes by creating migration guides in the `UPGRADING.md` file

- **Does this PR introduce a breaking change, and is [labeled accordingly](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)?**

No.

* **Other information**:

---

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

### CODE REVIEW

**Comment by silvanocerza** [comment]
What should be considered a breaking change though? I think we should define that to know exactly when to mark a certain PR with that label.

Also wouldn't it better if the label is an actual issue label instead of text "polluting" the PR title?

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (per1234) [inline .github/PULL_REQUEST_TEMPLATE.md:20]: I wonder it the term "labeled" is clear enough? It might lead people to think of GitHub's labeling feature.

I have two ideas of how things could be made more clear:
- Change the term to something like "titled".
- Add an anchor to the specific bullet point in the contributing guidelines, and link to that:
    ```md
    - <a id="breaking"></a>Please start the commit message and PR title with the string **[breaking]** if the PR contains a breaking change. A
    ```

File: docs/CONTRIBUTING.md
Comment (per1234) [inline docs/CONTRIBUTING.md:57]:
```suggestion
- If the PR contains a breaking change, please start the commit message and PR title with the string **[breaking]**. A
```
I think this wording improves readability because it provides the condition up front, so it's easier to understand the purpose of the bullet point when skimming the document. That makes it easier for those who need it to find, and faster for those who don't need it to skip over.

**Review by per1234 — CHANGES_REQUESTED** [review]

**Comment by per1234** [comment]
>Also wouldn't it better if the label is an actual issue label instead of text "polluting" the PR title?

I do sometimes get a bit annoyed by the "pollution" of the titles with these labels. It makes it more difficult to write a meaningful title that fits in the character limit. However, I think this "pollution" is worth its weight in precious characters.

The labeling is a great idea, but we must keep in mind that community contributors can't label their PRs. I haven't seen it with PRs, but I do see periodically see comments from people submitting issues who are concerned because they wanted to follow the convention they see on the other issues of adding labels, but can't figure out how to do it (because it's impossible for them). So it would need to be clearly communicated. It becomes more of a maintainer guideline at that point than a contributor guideline

**Comment by per1234** [comment]
I forgot to add that I investigated whether GitHub's pull request template system allows automatically labeling PRs according to  front matter metadata, like the issue template system does, but it seems this is not supported.

**Comment by obra** [comment]
Over on #1044, @rsora suggested that I chime in here.

In my eyes, a "breaking" change is a one that intentionally makes the user change their code in order to keep working.


I believe that any PR for a breaking change should include an entry in an "UPGRADING" file that's separate from the main Changelog. That entry should include instructions telling a smart (but possibly uneducated) user what they need to do differently, ideally with an example and a note about which version of the tool the functionality was removed in.

In a perfect world, breaking changes should include code to warn the user when they're using the old, deprecated functionality. Ideally, the warning message should be a concise version of the UPGRADING entry and should tell the user exactly what to change to make their old code go.

Maybe the question for this template to ask is something like:

"If this PR is merged, will any users need to change their code, command-line invocations, build scripts or data files when upgrading from an older version of `arduino-cli`?"

File: docs/CONTRIBUTING.md
Comment (rsora) [inline docs/CONTRIBUTING.md:57]: @silvanocerza 

> What should be considered a breaking change though?

I added a better description to to both the PR template and the docs borrowing words from @obra (Thanks :smile_cat: ). You like it?

@per1234 Nice suggestion about adding the anchor :+1: Please review the new wording like only you can do :detective:

**Comment by rsora** [comment]
DDoS countermeasures [cause CI to fai](https://github.com/arduino/arduino-cli/pull/1039/checks?check_run_id=1357766320#step:11:22)
I'll open a separate PR after a thorough arduino.cc links checks

File: docs/CONTRIBUTING.md
Comment (silvanocerza) [reply docs/CONTRIBUTING.md:57]: Yep, now it's clear what is meant.

**Comment by rsora** [comment]
@obra I very like your suggestion:

> I believe that any PR for a breaking change should include an entry in an "UPGRADING" file that's separate from the main Changelog.

As a user, where would you expect to see this file?
1. bundled with the binary?
2. as a Github Release attachment?

Having it as a section of the Release changelog we put in the release description could be easier for us to maintain and immediately available to users, as they can see it when they download the binary. OTOH we have other distribution channels for the CLI, in that case, maybe a file bundled with the release could reach the majority of our users. 
I would love to have any hint from you on this :smiley_cat:

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (per1234) [inline .github/PULL_REQUEST_TEMPLATE.md:22]:
```suggestion
when upgrading from an older version of Arduino CLI? -->
```
My understanding is that the name of the application is "Arduino CLI" and `arduino-cli` is only the command name. Similar to Git vs. `git`.

File: docs/CONTRIBUTING.md
Comment (per1234) [inline docs/CONTRIBUTING.md:60]:
```suggestion
- <a id="breaking"></a> If the PR contains a breaking change, please start the commit message and PR title with the
  string **[breaking]**. Don't forget to describe what changes users might need to make in their workflow or application
  due to this PR. A breaking change is a change that forces users to change their code, command-line invocations, build
  scripts or data files when upgrading from an older version of Arduino CLI.
```
I think this wording improves readability because it provides the condition up front, so it's easier to understand the purpose of the bullet point when skimming the document. That makes it easier for those who need it to find, and faster for those who don't need it to skip over.

---
I also made the change from `arduino-cli` -> "Arduino CLI" as in my suggestion above. so if I'm wrong about the semantics, that part of the change must be reverted.

**Review by per1234 — CHANGES_REQUESTED** [review]

**Review by per1234 — APPROVED** [review]
Great work @rsora!

**Comment by obra** [comment]
> As a user, where would you expect to see this file?
> 
> 1. bundled with the binary?
> 2. as a Github Release attachment?
> 
> Having it as a section of the Release changelog we put in the release description could be easier for us to maintain and immediately available to users, as they can see it when they download the binary from the release page. OTOH we have other distribution channels for the CLI, in that case, maybe a file bundled with the release could reach the majority of our users.
> I would love to have any hint from you on this 😺

Sorry for my slow reply.

One of the cool things about arduino-cli is that it's a self-contained executable and may well be distributed unpackaged.  Because of that, I don't think a bundled file names sense.

A pattern that's somewhat common is an "UPGRADING" file in the repository containing a running log of breaking changes, but especially for a project with nice docs, having it on the documentation website is also reasonable, https://arduino.github.io/arduino-cli/ seems like a pretty natural place to have an "Upgrading" section.

I think that putting the content on the GitHub releases page isn't great. That's where you go to get the release, not where to learn about it. It's also not a place a PR making a breaking change can add content to ;)  I'm a big fan of having the documentation about the breaking change be included in the PR that makes the change.

**Comment by rsora** [comment]
>especially for a project with nice docs, having it on the documentation website is also reasonable, https://arduino.github.io/arduino-cli/ seems like a pretty natural place to have an "Upgrading" section.

This is an excellent Idea @obra, this should be compatible with our release workflow and also with how our documentation release process works!

Having different documentation "versions" bound to each release and each release bound to a "release branch", for each release that contains breaking changes it should be sufficient to add the proper changes to the "upgrading" docs section as a commit on the related release branch.

@per1234 WDYT? It sounds to me that this could be the right tradeoff in terms of maintainability for the team.

**Comment by per1234** [comment]
>an "UPGRADING" file that's separate from the main Changelog.

I don't see any benefit to having separate documents for the changelog and the migration guide. This information is fundamentally related.

>A pattern that's somewhat common is an "UPGRADING" file in the repository containing a running log of breaking changes, but especially for a project with nice docs, having it on the documentation website is also reasonable,

The documentation content [is hosted in the repository](https://github.com/arduino/arduino-cli/tree/master/docs), so we would get both.

>It's also not a place a PR making a breaking change can add content to ;) I'm a big fan of having the documentation about the breaking change be included in the PR that makes the change.

I very much agree with this sentiment. I would suggest just having a link from the releases page to that release's section of the changelog documentation page.

>for each release that contains breaking changes it should be sufficient to add the proper changes to the "upgrading" docs section as a commit on the related release branch.

I think we should have a running document with a section for each release. The updates to the document should be committed to the `master` branch, then the release branch and the versioned documentation gets the snapshot of the file at that point in the history. This is how it's done in the Arduino IDE's repo:
https://github.com/arduino/Arduino/blob/master/build/shared/revisions.txt

The changelog and and any migration guides that weren't added to the changelog along with the PR that made the breaking change can be submitted as a PR to the repository before making the release.

**Comment by silvanocerza** [comment]
I talked with @rsora and we reached a conclusion for this. 

We updated our internal release process to handle the release containing breaking changes.

Also we updated the contributing guidelines with new steps that require the contributors to edit the `UPGRADING.md` file whenever a breaking change is introduced with a PR they personally create following the suggestions from @obra. Thanks!

This will help us keep track of breaking changes and the user to handle them accordingly when a new version is released.

@per1234 may I ask you for a new review? Also could you take a look to the newly updated internal release process?

For the record we are discussing internally to publish our release process guidelines.

Don't be bothered about the `verify-links` check failing, it's complaining that it can't find the new upgrading page on the documentation but that's normal until we merge this PR.

File: docs/CONTRIBUTING.md
Comment (per1234) [inline docs/CONTRIBUTING.md:429]:
```suggestion
[upgrading-file]: UPGRADING.md
```
This allows the link to work when viewing the documentation via the GitHub web interface, which is especially likely with CONTRIBUTING.md, since GitHub prompts contributors to read it there.

File: docs/UPGRADING.md
Comment (per1234) [inline docs/UPGRADING.md:5]: I don't think the linked information is of interest to the target reader of this file, so I would just remove this sentence.

If it is to be retained, please change the link to be relative:
```markdown
[contrib-docs]: CONTRIBUTING.md#breaking
```

File: mkdocs.yml
Comment (per1234) [inline mkdocs.yml:64]:
```suggestion
  - UPGRADING.md
  - getting-started.md
  - command-line-completion.md
  - CONTRIBUTING.md
```
It seems better placed in the index directly under "Installation".

**Review by per1234 — CHANGES_REQUESTED** [review]

File: docs/UPGRADING.md
Comment (silvanocerza) [reply docs/UPGRADING.md:5]: Right, didn't think about that. I'll just remove it.

File: docs/UPGRADING.md
Comment (per1234) [inline docs/UPGRADING.md:7]:
```suggestion
```
This is no longer needed.

**Review by per1234 — CHANGES_REQUESTED** [review]

File: docs/UPGRADING.md
Comment (silvanocerza) [reply docs/UPGRADING.md:7]: Right, I forgot. :facepalm:

**Review by per1234 — APPROVED** [review]

**Pull request merged**
